### ISSUE TITLE
Multi const string concat using interpolation - not optimized

### ISSUE THREAD

**Issue by TahirAhmadov** [dotnet/roslyn#61405]
This was posted on everybody's favorite blog, thedailywtf.com. 

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgMwAEOZAwmQN5FmMXmYAsZAsgBQCUdDTQZlwAGMgAsAphAAmkgE5kAvGQBEqgNwDGU2QuVqAKlAQAbSZu0TpcxSoAkq2rtsBfOAB1VAMSjyUCGQAchBgkl6WhIIu+g5OMfLuXhxQMjLmwaHh6lpRTAkGjs42CkmqADKogSFhEbnRJXZkRQnuAMoArgBmXVAAHpENek0tje5UAPZgAA4QMACeg/mNhfFjcAAikrPyCGEwCEs6K3HFw2UAUhPAZMZm2UfWw6tnbp6qAEIdKLCSKChkNoIeSSSQIOpWAqnVrvL4/GB/AFAkFgsjYCF5Y7PaHrLxw37/QHA0GBEgYoa2F4wvHfAkAqgmebk5bY5prc6w2kIwlAiAIB71FmUnEcmnwxFkAAKEwCEFM1AmcmZWOFbNepU54sJkw6B3k8wA9GQAEqSADmUAmMGVT1Vo1FqgAElNJESUeCcpCTmrqU6XW6SWibVCfbi/WEA6iyZ7MbbYqGHc6IwyEEyYxT4/a3l4k67efzg96sxqc/7pbL5ZMlemhZn2dnw66dXrDSbzZbrTWVXX1Yl3gB5BBSRTIkmF1nFvteQfDyOBdFduMjesl1Qz/SjqPju0rqdrof6FNpx4hydldcjhB8gVeie788HxTlq+VxU32On+8Dx8K3XA1umhaVrbj2voAIL/FAsoHAA5ACkriFa74ZsuvZlPi3IAt4EB9CBqG+hhEoIUheFUmGhGEsRCJBoun5oXAVByqYwAQAAxgA1ieRZfl4jFPohCKkSKDaTDMczzJw0AwFKAnIbW+FhrmZDYbhtHcfRpYRlRcndgpiZlrJNFcXe9EAJJtBsQTGTuGmqBw1xQBk2lCQmDYXspOEuWe36zs5akmTCkoQGaCjWaBYaSvIUCQPqMkkf5Nm+sasiWnFgkJeFDqGIYACaBqGBsGxpTpS5kRyhiSOYAyCrpZVvGBrGsRMf5hXp9UwDAUAAG4KCgED6q1dWrhBPzQQgcGZLUGVtauHyOaYsBmmQpkwF0EzyJACAdl5X5zbs4gyBAiw1aVwmzVyEpgWkIKUf2ZAfBMqmDWdfaMfyZrrVAfzPa5GpUOIjkyCCMA/d5XgbH4kisQg60SW0Cg9fIO22QAomQAC0ZCQI5ZBXUDiLI76aOY9j8qGPM0wlXRRMY1j0DyhDKDTKYR2TVT6lE+jpNorj10E9NQ17sTdM49gtwU+zAVhsL3Ni4zzOszUkuJdLXP02Q5B4zdKCE9LtPc+Q5OU7rDoy+r5DyyzElKybbwAOKSDAtigzxqh2xMiMwPsgSmUVQQdGAwAKLbGrOsAwDHbeKv6RGWtEXdD1PSd1PrCtJhQHKKAu7ZK38vICKBN4KL3d8x7Jxz6wANKSPMADu60yFn5dSxylQwGaHTBRYzfR285QTKxfIdtnvocHMXfyBNNsCy97gpOYXcj+sQQTPyTdR5lbk9FArGuv3g9bcBM+/Xu/byGacxQAAXkPVpystfsB0HSPH2D+6znHt33Y9Id9pFlpRVTEvDk/8urXmAW8SKEwuiImHj3Teq5TQwPkCCGQ90y4bxmn2eGMAfhbS6oyCBGo2jTGaigbumDBZlAAKrkMULgX+NC6FGXgVgph+hoxEL3LQ/QrBGHvAAOqSBuEFEKNohC4AAJxcASDwE6rgiCuCAA)

Is it possible to optimize this further to get rid of the multiple `Concat` calls? I know that code is atrocious (hence the source) but still, I can imagine more sane variations of this in "normal" codebases.

**Comment by CyrusNajmabadi** [comment]
I'm not a fan of issue raised that don't actually specify an actual problem, or show this is widespread enough or important enough to warrant a specialized optimization for this.   I'm guessing this is asking for some sort of analysis that some series of string operations continually operates on a single location (using it as an input and output) and somehow decides to change the emit strategy.

For this case, it seems like it would have to be super conservative (in case the value was observable in between assignments).

IMO, this would be better addressed with someone (including in the community) writing some sort of analyzer that looks for unnecessary string operations that cause lots of temp values that will be overwritten and which suggests updating it to something less wasteful.

**Comment by TahirAhmadov** [comment]
I'm not 100% sure at what stage this should be done: parsing, emittance, or JIT. My thought was, it should be able to take that interpolation code (in my original SharpLab) and produce a single string constant (like in @stephentoub  's SharpLab) when the program is actually run. You can see that even if you switch to JIT Asm, it's still a series of `Concat` calls.

**Comment by jcouv** [comment]
I don't think we should try to optimize the codegen for this specific/uncommon pattern (even the article says it's abusing string interpolations). It doesn't seem worthwhile.
